File: DesignPatterns.UI.Web/ClientApp/src/app/models/Adapter/PolicyFeeAdapter.ts
import { IPolicyFeeAdapter } from "./IPolicyFeeAdapter";
import { PolicyFee } from "./PolicyFee";

export class PolicyFeeAdapter implements IPolicyFeeAdapter
{
    private policyFee: PolicyFee;

    constructor(inputPolicy: PolicyFee, familyMembers: number) {
        this.policyFee = inputPolicy;

        this.PolicyHolderName = this.policyFee.PolicyHolderName;
        this.PolicyType = this.policyFee.PolicyType;
        this.BaseFee = this.policyFee.Fee;

        if(familyMembers < 1) familyMembers = 1;
        this.FamilyMembers = familyMembers;

        this.PolicyFeeDiscount = this.calculateDiscount();
        this.Fee = this.calculateFee();
    }

    PolicyHolderName: string = '';
    PolicyType: string = '';
    BaseFee: number = 0;

    FamilyMembers: number = 0;
    PolicyFeeDiscount: number = 0;

    Fee: number = 0;

    private calculateDiscount() {
        if(this.FamilyMembers < 2) return 0;

        switch(this.FamilyMembers) {
            case 2:
                {
                    return 5;
                }
            case 3:
                {
                    return 7.5;
                }
            case 4:
                {
                    return 10;
                }
            case 5:
                {
                    return 12.5;
                }
            default:
                {
                    return 15;
                }
        }
    }

    private calculateFee() {
        return (this.BaseFee * this.FamilyMembers) * (100 - this.PolicyFeeDiscount) / 100;
    }
}

File: DesignPatterns.AdapterPattern/Models/PolicyFeeAdapter.cs
using DesignPatterns.AdapterPattern.Interfaces;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DesignPatterns.AdapterPattern.Models
{
    public class PolicyFeeAdapter : IPolicyFeeAdapter
    {
        private readonly PolicyFee _individualPolicyFee;

        public PolicyFeeAdapter(PolicyFee individualPolicyFee)
        {
            _individualPolicyFee = individualPolicyFee;
        }

        public int FamilyMembers { get; set; }

        public decimal PolicyFeeDiscount
        {
            get
            {
                if(FamilyMembers < 2) return 0;

                switch (FamilyMembers)
                {
                    case 2:
                        {
                            return 5m;
                        }
                    case 3:
                        {
                            return 7.5m;
                        }
                    case 4:
                        {
                            return 10m;
                        }
                    default:
                        {
                            return 20m;
                        }
                }
            }
        }

        public decimal PolicyFeeDiscountValue
        {
            get
            {
                return BaseFee * PolicyFeeDiscount / 100;
            }
        }

        public string PolicyHolderName
        {
            get
            {
                return _individualPolicyFee.PolicyHolderName;
            }
        }

        public string PolicyType
        {
            get
            {
                return _individualPolicyFee.PolicyType;
            }
        }

        public decimal BaseFee
        {
            get
            {
                return _individualPolicyFee.Fee * FamilyMembers;
            }
        }
        
        public decimal Fee
        {
            get
            {
                return BaseFee - PolicyFeeDiscountValue;
            }
        }
    }
}
